### ISSUE TITLE
client: implement Is() and Unwrap() in status.Error to not mask context.Canceled

### ISSUE THREAD

**Issue by pracucci** [grpc/grpc-go#6862]
### Use case(s) - what problem will this feature solve?

The gRPC client (e.g. `grpc.ClientStream`) returns a gRPC error when the context is canceled or the deadline is exceeded. The actual standard golang error (e.g. `context.Canceled`) is masked by the gRPC error and so `errors.Is(grpcErr, context.Canceled)` returns `false` even if the error has been actually generated by context being canceled.

This behaviour makes checking for context cancellation (or deadline exceed) a bit cumbersome in our application because we have to re-map the returned gRPC error into a `context.Canceled` or `context.DeadlineExceeded`, given the rest of the application code is already instrumented to check it but not gRPC-specific error codes. In other words, what I mean is that we're used to check gRPC-specific error codes in the code blocks directly interacting with gRPC but when these errors are returned by functions, the callers are typically gRPC agnostic.

### Proposed Solution

I propose to implement `Is(err error) bool` and `Unwrap() error` on `status.Error`:

- `status.Error.Is(context.Canceled)` would return `true` if the status error is `codes.Canceled()` (and similar for `context.DeadlineExceeded`
- `status.Error.Unwrap()` would return `context.Canceled` if the status error is `codes.Canceled()` (and similar for `context.DeadlineExceeded`

_Does it make sense or are there good reasons why we shouldn't do it? If we'll get a consensus (either on this proposal or a different one after the discussion) I'm open to work on the code changes._

**Comment by dfawley** [comment]
Interesting idea.  There is one hard case to consider:

* `CANCELED` and `DEADLINE_EXCEEDED` are status codes that can be returned arbitrarily from the server even if the client didn't cancel the RPC or reach its deadline.  Is it proper to return `context` errors (or something that `Is()` them) in that case?

> when these errors are returned by functions, the callers are typically gRPC agnostic.

In these cases, wouldn't the callers usually expect to be able to do a direct comparison to `context.Canceled` vs. an `errors.Is()` call?

If you're making a library that wraps gRPC in an attempt to hide gRPC from your users, then you should really consider doing this conversion (and converting all other errors as well) before returning them.  If you directly pass through our return values, our API becomes part of your API.

**Comment by pracucci** [comment]
> CANCELED and DEADLINE_EXCEEDED are status codes that can be returned arbitrarily from the server even if the client didn't cancel the RPC or reach its deadline. Is it proper to return context errors (or something that Is() them) in that case?

I was expecting this comment and I'm glad you raised the point. I don't have enough knowledge about gRPC status contract to have a strong opinion whether it's proper or not to return an error which `Is()` a `context.Canceled` (or `context.DeadlineExceeded`) is such case, and I would defer to your experience. If it's not, then I still believe we could find a way to propagate whether it's a context canceled/exceeded within the `status.Error`.

> In these cases, wouldn't the callers usually expect to be able to do a direct comparison to context.Canceled vs. an errors.Is() call?

Direct comparison has a limitation: doesn't work with wrapped errors. In our application we're used to wrap errors, so it's typical to do something like `errors.Wrap(err, "fetching xxx from yyy")`. If `err` is a `context.Canceled`, the wrapped error would fail a direct comparison but would still succeed a `errors.Is(err, context.Canceled)`.

> If you're making a library that wraps gRPC in an attempt to hide gRPC from your users, then you should really consider doing this conversion (and converting all other errors as well) before returning them. If you directly pass through our return values, our API becomes part of your API.

Until today we were passing around gRPC errors, but we never realised that context canceled errors were mapped as `status.Error` (I know, we've realised it late). To stop the bleed we've started wrapping gRPC client to a custom one which does a remapping of errors, but there's quite a lot of boilerplate to write to wrap all clients.

For this reason I'm wondering if adding `Is()` to `status.Error` (this proposal) could be a viable approach for the wide gRPC golang community. If not, we'll keep wrapping gRPC clients (I'm thinking to write a protoc plugin to do it), otherwise I could invest my time to do the same in grpc-go.

**Comment by dfawley** [comment]
After thinking about this more...

1. I think it would not be appropriate to wrap `context.Canceled`/`DeadlineExceeded` if the server returns those statuses, because the context itself may still be valid.
2. We could possibly (? -it might be easy, or it might be very very hard) detect whether the cancelation/deadline happened on the client-side and only conditionally wrap.
3. If we did 2, it's still theoretically possible there could be a race between the server deciding the client's deadline was exceeded and the client noticing the context expired, meaning we could get it wrong sometimes.

tl;dr I think I'd rather keep things simple here and not worry about all this complexity.  That said, people have asked for similar things, e.g. #3328, so maybe this is something we'd consider in the future.

**Comment by pracucci** [comment]
Ok. I see your point. Then we'll wrap the gRPC client and address it on our side. Thanks for your time.

**Comment by pstibrany** [comment]
@dfawley @pracucci  What do you think about adding an option to gRPC client to use custom error mapper function, instead of the default one? I think that might be enough for our needs, and it would save us from wrapping all the clients.

**Comment by nbgraham** [comment]
In the referenced [error mapper function](https://github.com/grpc/grpc-go/blob/fe1519ecf7aee0f93ece33be8a682d80d24a95c8/rpc_util.go#L873-L898), I don't understand how masking these errors is helpful.

>If we did 2, it's still theoretically possible there could be a race between the server deciding the client's deadline was exceeded and the client noticing the context expired, meaning we could get it wrong sometimes.

Even if it was wrong, isn't having the same `Error()` string equally confusing?

How is this
https://github.com/grpc/grpc-go/blob/fe1519ecf7aee0f93ece33be8a682d80d24a95c8/rpc_util.go#L870
better than this
```go
  status.fromError(codes.Canceled, context.Canceled) // something that implements Unwrap
```
?

It's very confusing the see the output from `Error()` _look_ identical to something that would `Unwrap()` into `context.Canceled` but actually doesn't.

**Comment by dfawley** [comment]
> I don't understand how masking these errors is helpful.

Per https://pkg.go.dev/google.golang.org/grpc#ClientStream: "All errors returned from ClientStream methods are compatible with the status package."

This isn't "masking", it's ensuring all the errors are of a consistent type.  Even if we did think it'd be better to return different types, we wouldn't be able to change the behavior, as it would be a breaking change.

Re @pstibrany's suggestion:
> an option to gRPC client to use custom error mapper function

The function linked is not some general-purpose, well-defined interface.  It coerces errors from different sources into that uniform `status` type.  One of those sources is our (internal) transport, which no external program would be able to handle.

I'm not sure, but it _might_ be possible to do something like:

- Make the `ClientStream` functions return `context.Canceled`/`DeadlineExceeded` _directly_, when appropriate.
- Add support for these errors to the `status` package (i.e. `FromError` can convert from `context.*` into `status.Status`).

Then, technically, `ClientStream` still only returns errors "compatible with the status package", but you could then also do:

```go
if res, err := client.Call(...); err == context.DeadlineExceeded {
...
```

In theory this still races, though, as the server's deadline could elapse slightly before the client's, or the client could be under heavy load and miss the context cancellation.  In these situations, the client could receive the status on the wire from the server and return a `status.Error` error instead of the `context.DeadlineExceeded`.  This could then cause the client code to make wrong assumptions.  (Maybe the consequences of those assumptions would always be insignificant, though?)

Either way, this issue proposed a very specific approach, which I don't think we want to pursue.  If you want to discuss this topic further, please file a new issue.
